keps/sig-api-machinery/5073-declarative-validation-with-validation-gen/README.md

```diff
@@ -0,0 +1,1266 @@
+                        return
+                }(&obj.MinReadySeconds, safe.Field(oldObj, func(oldObj *corev1.ReplicationControllerSpec) *int32 { return &oldObj.MinReadySeconds }), fldPath.Child("minReadySeconds"))...)
+        return errs
+}
+```
+
+This generated code will then be used in the kube-apiserver to validate API requests.
+
+## Motivation
+
+Kubernetes API validation rules are currently written by hand, which makes them difficult for users to access directly, review, maintain, and test.
+
+Declarative validation will benefit Kubernetes maintainers:
+
+*   It will make it easier to develop, maintain and review APIs.
+*   It will make it easier programmatically inspect and analyze the API, enabling new tools and improved documentation.
+*   It will enable improvements to the API machinery. For example, a feature like ratcheting validation will become more tractable to implement because the feature can be implemented once in the declarative validation subsystem rather than piecemeal across the 15k lines of hand written validation code.
+
+Declarative validation will also benefit Kubernetes users:
+
+*   Adding new fields and associated validation rules becomes a simpler process of adding IDL tags to the field definition, rather than writing and maintaining validation functions. This reduces potential inconsistencies and bugs.  Testing is also streamlined as centralized rules and frameworks enable the use of test fixtures across k8s types making creating tests simpler and faster to implement for contributors.
+*   Creating k8s APIs becomes faster as developers can focus on the API’s structure and behavior, leaving the validation boilerplate to be generated automatically. This speeds up development and encourages experimentation with new APIs.
+*   Validation is performed on versioned types (assuming recommended approach is used from design below), so error message and attributed field path is more likely to be relevant to the user. (ex: some fields in autoscaling/v2 will be mis-identified with current approach but resolved w/ Declarative Validation + versioned types)
+
+Additionally Declarative Validation can be built upon to support features such as:
+
+*   Ability to give users direct access to the actual API validation rules (via publishing OpenAPI rules for native types), which are currently only available to developers willing and able to find and read the hand written validation rules.
+*   Ability to enable clients to perform validation of native types earlier in development workflows ("shift-left"), such as with a Git pre-submit linter.
+*   Ability to improve API composition. In particular CRDs that embed native types (such as PodTemplate), which gain validation of the native type automatically. This has the potential to simplify controller development and improve end user experiences when using CRDs.
+*   (Eventually) bring CRD validation and built-in validation together with the same set of server-side validations.
```

Comment:
I could be wrong on this, but I thought there was an overall architectural goal to push a large portion of new API development out of k/k and into CRDs. If I'm understanding that correctly, why aren't CRDs an immediate goal?

Reply:
The most immediate goal is to lessen the burden and risk of hand-written validation code and to simultaneously lay pavement for CRDs to call the same code, with ~0 perf hit.